### DISCUSSION TITLE
Service container or DI container?

### DISCUSSION THREAD

**Discussion by ghost** [symfony/symfony#44832] [Q&A]
I'm opening this discussion because I've recently dived into the services, containers, stateless/-ful considerations.

As far as I can see, terms Dependency Injection Container and Service Container are used interchangeably. But are both really equal? Service Container, as even the name suggests, contains services. Services are objects that "do something", they have behavior. They also should be stateless (except very few specific use cases). On the other hand Dependency Injection is a pattern where objects have their dependencies injected via constructor and something else is responsible for creating them. This "something" is a DI Container that is a factory for the objects and their dependencies. 

What I am wondering about is: **if my service depends on an object that is not a service, like data object or value object (e.g. some Configuration object which stores only some fixed configuration values) what should I do?** It's not a service so I can't add it to the Service Container but it is a dependency of my service so somehow must be inside DI Container so it can inject this to my service.

Other interesting thing is this data object versus value object thing. Value object are immutable so even if I inject them as dependency to my service then this service will not become stateful as any modification to the VO does not change the original object. However this is not the case with (mutable) data objects.

I wonder is someone with more experience can elaborate a little on the topic and share their knowledge.

**Comment by ro0NL** [comment]
Strictly spoken it's a DI container: https://www.php-fig.org/psr/psr-11/

Not sure what the actual question is, but i consider it a giant object registry+factory yes (the root of all things).

Symfony demotes using it direclty (thus as a service locator): https://symfony.com/doc/current/service_container.html#public-versus-private-services

you can create dedicated service locators derived from the DI container: https://symfony.com/doc/current/service_container/service_subscribers_locators.html#defining-a-service-locator

**Reply by ghost** [reply]
I've bolded the question.

**Reply by ro0NL** [reply]
Right. I tried to explain the same as @stof did, that is you can put anything in the DI container. It's a factory DSL in that sense.

Calling it a service container is maybe confusing, but that's mostly a semantical (non)issue.

**Comment by stof** [comment]
Configuration objects needed by a service can be defined as services in the container too (Symfony itself does it a lot).

Value objects that should not go in the container are objects for which you want to use multiple instances to operate on in your services (but those cannot be passed as constructor argument in services either, otherwise you would have to recreate the service to operate on them)
